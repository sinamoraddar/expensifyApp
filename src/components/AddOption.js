import React from 'react';
export default class AddOption extends React.Component {
  state = {
    error:undefined,
  }  
    handleAddOption = (e) => {
      e.preventDefault();
      const optionInput = e.target.elements.optionInput.value.trim();
      const error = this.props.handleAddOption(optionInput);
      this.setState(()=>({error}));
      if(!error){
        e.target.elements.optionInput.value = '';
      }
    };
    render(){
      return (
        <div>
          {this.state.error && <p>{this.state.error}</p>}
          <form onSubmit = {this.handleAddOption}>
            <input type="text" name="optionInput" id="optionInput"/>
            <button>
              Add Option
            </button>
          </form>
        </div>
      );
    }
  }
 
  