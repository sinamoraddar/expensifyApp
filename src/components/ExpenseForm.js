import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { addExpense } from '../actions/expenses';

// const date = new Date();

// const now = moment();
// console.log(now.format("Do MMM YYYY"))

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({
            description
        }));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({
            note
        }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({
                amount
            }))
        }

    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({
                createdAt
            }))
        }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }))
    }
    onSubmitForm = (e) => {
        e.preventDefault();
        const { description, amount, note, createdAt } = this.state;
        const { onSubmit } = this.props;
        if (!description || !amount) {
            this.setState(() => ({
                error: 'please provide description or amount'
            }))
        } else {
            this.setState(() => ({
                error: ''
            }))
            onSubmit({
                description,
                amount: parseFloat(amount) * 100,
                note,
                createdAt: createdAt.valueOf()
            })
        }

    }

    render() {
        return (
            <form className="form" onSubmit={this.onSubmitForm}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    className="text-input"
                    type="text"
                    name="description"
                    placeholder="Description"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    autoFocus
                />
                <input
                    className="text-input"
                    type="text"
                    name="amount"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    className="textarea"
                    name="note"
                    placeholder="Add a note for your expense here (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}>
                </textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        )
    }
};
export default ExpenseForm;

//  {/* note desc createdAt amount */}
//  <form onSubmit={(e) => {
//     e.preventDefault();
//     dispatch(addExpense({
//         description: e.target.description.value,
//         amount: e.target.amount.value,
//         note: e.target.note.value,
//         createdAt: e.target.createdAt.value
//     }))
// }}>
//     <input placeholder="description" type="text" name="description" />
//     <input placeholder="amount" type="number" name="amount" />
//     <input placeholder="note" type="text" name="note" />
//     <input placeholder="created at" type="number" name="createdAt" />
//     <button>Create</button>
// </form>