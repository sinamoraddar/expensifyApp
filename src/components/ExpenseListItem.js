import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <div>
        <ul>
            <Link to={`/edit/${id}`}><li>description: {description}</li></Link>
            <li>amount: {amount}</li>
            <li>created at(milliseconds from 1/1/1970): {createdAt}</li>
        </ul>
    </div>
);

export default ExpenseListItem;