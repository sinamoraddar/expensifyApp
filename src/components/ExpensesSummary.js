import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getVisibleExpenses from '../selectors/expenses';
import getTotalExpenses from '../selectors/expensesTotal';

export const ExpensesSummary = ({ expenseCount = 0, expensesTotal = 0 }) => {
    const expenseWord = (expenseCount === 1) ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div>
            <h3>
                Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
            </h3>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getTotalExpenses(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);