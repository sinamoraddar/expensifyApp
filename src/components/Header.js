import React from 'react';
import { NavLink } from "react-router-dom";
import { connect } from 'react-redux';
import { startLogOut } from '../actions/auth';

export const Header = ({ startLogOut }) => (
    <header>
        <h1>Expensify</h1>
        <NavLink exact to="/" activeClassName="is-active">Home</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
        <button onClick={startLogOut}>Log Out</button>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogOut: () => dispatch(startLogOut())
})

export default connect(undefined, mapDispatchToProps)(Header);