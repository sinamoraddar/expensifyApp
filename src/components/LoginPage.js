import React from 'react';
import { connect } from 'react-redux';
import { startLogIn } from '../actions/auth';

export const LoginPage = ({ startLogIn }) => (
    <div className="box-layout">
        <div className="box-layout__box">
        <h1 className="box-layout__title">Expensify</h1>
        <p>Always keep track of your expenses here !</p>
        <button className="button" onClick={startLogIn}>Login with Google</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogIn: () => dispatch(startLogIn())
})

export default connect(undefined, mapDispatchToProps)(LoginPage);