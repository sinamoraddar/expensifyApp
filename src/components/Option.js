import React from 'react';
const Option = (props) => (
      <div>
        <p>Option: {props.optionText}</p>
        <button 
          onClick={()=>{
            props.handleDeleteOption(props.optionText)
          }}
        >
          Remove
        </button>
      </div>
      );
  
  export default Option;