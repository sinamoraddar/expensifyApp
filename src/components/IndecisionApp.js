import React from 'react';
import Header from './Header';
import Action from './Action';
import Options from './Options';
import AddOption from './AddOption';
import OptionModal from './OptionModal';

class IndecisionApp extends React.Component {
    state={
        subtitle : 'Put your life in the hands of a computer',
        options:[],
        selectedOption:undefined
    }
    handleDeleteOption = (option) => {
      this.setState((prevState)=>({
          options: prevState.options.filter((item) => (
            item!=option
          )),
      }))
    };
    handleDeleteOptions = () => {
      this.setState(()=>({options:[]}));
    };
    handlePick = () =>  {
      const randomNumber = Math.floor(Math.random()*this.state.options.length);
      const randomOption = this.state.options[randomNumber];
      this.setState(() => ({
          selectedOption:randomOption
      }));
    };
    handleAddOption = (option) => {
      if(!option){
        return `Enter valid value to add item`;
      }else if(this.state.options.indexOf(option)>-1){
        return `This item already exists`
      }
      this.setState((prevState)=>({options:prevState.options.concat(option)}));
    };
    handleModalButton = () => {
        this.setState(() => ({
            selectedOption:undefined
        }));
    };
    componentDidMount(){
    try {
        const json = localStorage.getItem('options');
        const options = JSON.parse(json);
        if(options){
        this.setState(() => (
            {
                options
            }
            )
        );
        }
    } 
    catch (error) {
        //do nothing at all
    }
    
    
    }
    componentDidUpdate(prevProps, prevState){
    if(prevState.options.length!==this.state.options.length)
    {
        const json = JSON.stringify(this.state.options);
        localStorage.setItem('options',json);
    }

    }
    componentWillUnmount () {
    alert(`componentWillUnmount`)
    }
    render () {
      return (
        <div>
          <Header 
          subtitle={this.state.subtitle} />
          <Action 
          handlePick={this.handlePick} 
          hasOptions={this.state.options.length>0} 
          />
          <Options 
          options={this.state.options}
          handleDeleteOptions={this.handleDeleteOptions}
          handleDeleteOption={this.handleDeleteOption}
          />
          <AddOption 
          handleAddOption={this.handleAddOption} 
          />
          <OptionModal handleModalButton={this.handleModalButton} selectedOption={this.state.selectedOption} />
        </div>
      );
    }
  }
  export default IndecisionApp ;