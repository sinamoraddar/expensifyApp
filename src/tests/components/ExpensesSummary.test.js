import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should show the summary of one item in the list', () => {
    const list = {
        expenseCount: 1,
        expensesTotal: 1002
    }
    const wrapper = shallow(<ExpensesSummary {...list} />);
    expect(wrapper).toMatchSnapshot();

})
test('should show the summary of multiple expenses in the list', () => {
    const list = {
        expenseCount: 10,
        expensesTotal: 15000
    }
    const wrapper = shallow(<ExpensesSummary {...list} />);
    expect(wrapper).toMatchSnapshot();

})