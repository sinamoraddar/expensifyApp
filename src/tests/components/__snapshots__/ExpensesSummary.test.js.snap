// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show the summary of multiple expenses in the list 1`] = `
<div>
  <h3>
    Viewing 
    10
     
    expenses
     totalling 
    $150.00
  </h3>
</div>
`;

exports[`should show the summary of one item in the list 1`] = `
<div>
  <h3>
    Viewing 
    1
     
    expense
     totalling 
    $10.02
  </h3>
</div>
`;
